README.md
# 异常处理与错误码推荐的处理方式

## 1. 背景

曾经，我去面试过一家公司，面试官问我这样的问题：”你觉得在开发过程中，异常处理的代码大概占全部代码的多少比例？”当时的我可能还是涉世未深、对异常处理的理解没有那么深刻，回答的是“30%左右吧”。面试官当即就笑了：“小伙子，你还太年轻了，你说的可能要调转一下，正常的业务代码可能就占30%左右，其他的几乎全是异常处理。”

后面经历了多家公司的洗礼后，发现诚如当年那位大哥所说，异常处理的代码的确占据了绝大部分。这里需要提一下"二八原则"
，业务逻辑通常来说是没有太大难度的，可能就占据全部代码的20%左右，其他的80%都是用来处理异常、流量控制等提升程序健壮性、稳定性的代码。可以说异常处理是必不可少的，也是极其重要的。所以代码中就会出现大量的 try {...} catch(
...) {...} finally {...} 代码块，而且很多这样的异常处理逻辑都是相似的。这样不仅会导致大量的代码冗余，而且也会影响的代码的可读性。

请看如下两种风格的代码：

优化前的代码：

```Java
@ApiOperation(value = "登录以后返回token")
@RequestMapping(value = "/login", method = RequestMethod.POST)
@ResponseBody
public R login(@Validated @RequestBody UserLoginParam userLoginParam){
    try{
        String token=userService.login(userLoginParam.getUsername(), userLoginParam.getPassword());
        Map<String, String> tokenMap = new HashMap<>();
        tokenMap.put("token",token);
        return R.success(tokenMap);
    } catch (BusinessException e){
        log.warn("登录异常:{}", e.getMessage());
        return R.failed("登录异常:" + e.getMessage());
    } catch (Exception e){
        log.warn("登录异常:{}", e.getMessage());
        return R.failed("登录异常:" + e.getMessage());
    }
    return R.failed("登录异常");
}
```

优化后的代码：

```Java
@ApiOperation(value = "登录以后返回token")
@RequestMapping(value = "/login", method = RequestMethod.POST)
@ResponseBody
public R login(@Validated @RequestBody UserLoginParam userLoginParam){
    String token=userService.login(userLoginParam.getUsername(), userLoginParam.getPassword());
    Map<String, String> tokenMap = new HashMap<>();
    tokenMap.put("token",token);
    return R.success(tokenMap);
}
```

这两种风格的代码，我相信大部分的人都倾向于优化后的代码吧。可以看到，优化后的代码只处理了核心的业务逻辑，那么异常处理这一块的代码去哪里了呢？

## 2. 统一异常处理

回答上面的问题，就要提到Spring 3.2版本之后加入的注解：`@ControllerAdvice`，请看该注解的官方说明（截取一部分）：

```Java
/**
 * Specialization of @Component for classes that declare @ExceptionHandler, @InitBinder, 
 * or @ModelAttribute methods to be shared across multiple @Controller classes.
 * ...
 * By default, the methods in an @ControllerAdvice apply globally to all controllers. 
 */
```

翻译过来就是：

```Java
/**
 * 对声明了 @ExceptionHandler、@InitBinder 或 @ModelAttribute 方法的类的 @Component 进行特殊处理，
 * 将这些方法在多个 @Controller 类之间共享。
 * ...
 * 默认情况下，@ControllerAdvice 中的添加了以上注解的方法会应用于全局所有的控制器。
 */
```

我们再来看 `@ExceptionHandler` 注解的说明文档（截取一部分）：

```Java
/**
 * Annotation for handling exceptions in specific handler classes and/or handler methods.
 * Handler methods which are annotated with this annotation are allowed to have very flexible signatures. 
 * They may have parameters of the following types, in arbitrary order:
 * - An exception argument: declared as a general Exception or as a more specific exception. 
 *      This also serves as a mapping hint if the annotation itself does not narrow the exception types 
 *      through its value().
 * The following return types are supported for handler methods:
 * - @ResponseBody annotated methods (Servlet-only) to set the response content. 
 *      The return value will be converted to the response stream using message converters.
 */
```

翻译过来就是：

```Java
/**
 * 用于处理特定处理类和/或方法中的异常的注解。
 * 使用此注解注解的处理方法允许具有非常灵活的方法签名。
 * 它们可能具有以下类型的参数：
 * - Exception 参数：声明为一般异常或更具体的异常。
 *      如果注解本身没有通过它的 value 参数来缩小异常类型，这也可以作为一个映射提示。
 * 处理方法支持以下返回类型：
 * - 使用注解 @ResponseBody 注解的方法（仅限 Servlet），以设置响应内容。返回值将使用消息转换器转换为响应流。
 */
```

从以上的说明，我们可以知道，如果在程序中定义一个类，并将此类添加上 `@ControllerAdvice`
注解，然后在此类中添加一个（或多个）方法，方法的参数为想要处理的异常类，方法的返回值会需要返回的响应报文，并在方法上添加上 `@ExceptionHandler` 注解和 `@ResponseBody`
注解，那么我们就可以实现异常的统一处理了。

### 2.1 实战

上面讲的是理论，接下来进行实战。

#### 2.1.1 无统一异常处理 

1. 添加`BusinessException`

```java
public class BusinessException extends RuntimeException {
    // 此处为节省篇幅省略构造方法定义，继承RuntimeException的构造方法即可
}
```

2. 添加`UserService`接口

```java
public interface UserService {
    boolean login(String username, String password);
}
```

3. 创建`UserServiceImpl`，实现`UserService`接口

```java
@Service
public class UserServiceImpl implements UserService {

    @Override
    public boolean login(String username, String password) {
        if (username == null) {
            throw new BusinessException("User name cannot be null");
        }
        if (password == null) {
            throw new BusinessException("Password cannot be null");
        }
        // 此处只做演示作用，只允许admin登录，密码为123456
        if (!"admin".equals(username) || !"123456".equals(password)) {
            throw new BusinessException("User name or password is incorrect");
        }
        return true;
    }
}
```

4. 添加Controller

```java
@RestController
@RequestMapping("/user")
public class UserController {

    @Autowired
    private UserService userService;

    @Operation(summary = "Login")
    @GetMapping(value = "/login")
    public String login(@RequestParam(value = "username") String username,
                        @RequestParam(value = "password") String password) {
        boolean result = userService.login(username, password);
        if (result) {
            return "success";
        }
        return "failure";
    }

}
```

##### 2.1.1.1 测试

1. 正常使用admin/123456登录，应能成功

```shell
curl -X 'GET' 'http://localhost:8080/user/login?username=admin&password=123456' -H 'accept: */*'
```

返回：`success`

2. 使用错误的用户名和密码登录，会抛出`BusinessException`异常

```shell
curl -X 'GET' 'http://localhost:8080/user/login?username=wronguser&password=123456' -H 'accept: */*'
```

返回信息为：

```json
{
  "timestamp": "2022-04-19T06:59:34.556+00:00",
  "status": 500,
  "error": "Internal Server Error",
  "trace": "com.soulcraft.demo.errorhandling.demo1.exception.BusinessException: User name or password is incorrect...."
}
```

##### 2.1.1.2 小结

此处返回的错误信息包含了堆栈信息，对前端是十分不友好的。

具体代码示例可参见：[demo1](https://gitee.com/soulcraft/error-handling-demo/tree/main/demo1)

#### 2.1.2 异常处理进行统一处理

接下来，我们就为这个程序添加统一的异常处理。

`Controller`, `Service`实现与上面完全一样，只添加一个`UnifiedExceptionHandler`，统一处理异常。

注意：在此类上添加了`@ControllerAdvice`注解。

```java
@Slf4j
@ControllerAdvice
public class UnifiedExceptionHandler {

    @ExceptionHandler(value = BusinessException.class)
    @ResponseBody
    public String handleBusinessException(BusinessException e) {
        log.error(e.getMessage(), e);
        return "failure: " + e.getLocalizedMessage();
    }
}
```

##### 2.1.2.1 测试

使用错误的用户名和密码登录，不再抛出`BusinessException`异常

```shell
curl -X 'GET' 'http://localhost:8080/user/login?username=wronguser&password=123456' -H 'accept: */*'
```

返回信息为：

```text
failure: User name or password is incorrect
```
##### 2.1.2.2 小结

由此可见，只需添加一个统一的异常处理类即可将所有`Controller`中抛出的所有`BusinessException`统一到指定的方法中处理了。

既简单，又方便，代码又美观！何乐而不为？！

具体代码示例可参见：[demo2](https://gitee.com/soulcraft/error-handling-demo/tree/main/demo2)

## 3. 使用Asserts断言

我们发现，在`Service`中，有许多 `if (xx == null) {}` 这样的判断逻辑存在，这样的代码块是否可以继续优化呢？

```java
public boolean login(String username, String password) {
    if (username == null) {
        throw new BusinessException("User name cannot be null");
    }
    if (password == null) {
        throw new BusinessException("Password cannot be null");
    }
    // 此处只做演示作用，只允许admin登录，密码为123456
    if (!"admin".equals(username) || !"123456".equals(password)) {
        throw new BusinessException("User name or password is incorrect");
    }
    return true;
}
```

让我们再想想，我们是否可以参照 `JUnit` 框架中的 `Assertions` 类的处理？即对程序逻辑进行断言，如果断言不成立，则抛出异常，如果断言成立，则程序继续运行下一行代码。

优化后的代码是这样：

```java
public boolean login(String username, String password) {
    Asserts.assertNotEmpty(username, "User name cannot be null or empty");
    Asserts.assertNotEmpty(password, "Password cannot be null or empty");
    // 此处只做演示作用，只允许admin登录，密码为123456
    if (!"admin".equals(username) || !"123456".equals(password)) {
        throw new BusinessException("User name or password is incorrect");
    }
    return true;
}
```

在 `Asserts.assertNotEmpty` 中，如果检查参数为`null`或者字符串长度为空，则抛出 `BusinessException`。

```java
public final class Asserts {

    public static void assertNotEmpty(String obj, String message) {
        if (obj == null || obj.isEmpty()) {
            throw new BusinessException(message);
        }
    }
}
```

这样代码又清爽很多。

### 3.1 测试

使用空的用户名和非空密码登录

```shell
curl -X 'GET' 'http://localhost:8080/user/login?username=&password=123456' -H 'accept: */*'
```

返回信息为：

```text
failure: User name cannot be null or empty
```

测试通过。

具体代码示例可参见：[demo3](https://gitee.com/soulcraft/error-handling-demo/tree/main/demo3)

## 4. 不同异常的处理方式

按照上面的方法，虽然代码清爽了很多，但是所有的异常都抛出 `BusinessException`，这样子很不好区分到底后台是发生了什么样的错误，前端需要如何进行处理。

### 4.1 创建不同的异常类

我们可以针对不同的异常场景，创建不同的异常类。

```java
public boolean login(String username, String password) {
    if (username == null) {
        throw new InvalidParameterException("User name cannot be null");
    }
    if (password == null) {
        throw new InvalidParameterException("Password cannot be null");
    }
    // 此处只做演示作用，只允许admin登录，密码为123456
    if (!"admin".equals(username) || !"123456".equals(password)) {
        throw new UserLoginException("User name or password is incorrect");
    }
    return true;
}
```

上面这段代码我们需要新建两个异常类：`InvalidParameterException`，`UserLoginException`，都继承自`BusinessException`。这样的话，我们虽然解决了区分不同异常的问题，但是每种不同的场景，我们就需要新增一个异常类，就会造成程序中有许多异常类。并且每个异常对应的错误码也没有定义。这样的方法明显不是一个很好的方法。

那么应该如何解决这个问题，并且给每个异常加上不同的错误码呢？

### 4.2 期望的效果

我们想要的效果应该是，`Service` 代码如下所示：

```java
public boolean login(String username, String password) {
    UserResponse.USERNAME_CANNOT_BE_EMPTY.assertStringNotEmpty(username);
    UserResponse.PASSWORD_CANNOT_BE_EMPTY.assertStringNotEmpty(password);
    if (!"admin".equals(username) || !"123456".equals(password)) {
        UserResponse.USER_LOGIN_FAILED.throwNewException();
    }
    return true;
}
```

1. 当用户名为空时，前端收到的返回信息为：

```json
{"code":600,"message":"Username cannot be null or empty"}
```

2. 当用户名不为空、密码为空时，前端收到的返回信息为：

```json
{"code":601,"message":"Password cannot be null or empty"}
```

3. 当用户名不为空、密码不为空，但是用户名或密码不正确时，前端收到的返回信息为：

```json
{"code":602,"message":"User login failed"}
```

### 4.3 如何实现

那么应该如何实现如上描述的效果呢？

1. 首先，我们定义一个响应报应的接口`IResponse`，它只包括两个元素，`int`类型的`code`（错误码）以及`String`类型的`message`（错误信息）：

```java
public interface IResponse {
    int getCode();
    String getMessage();
}
```

2. 定义一个基础的实现类`BaseResponse`：

```java
@Getter
@Setter
public abstract class BaseResponse implements IResponse {
    private int code;
    private String message;

    public BaseResponse(int code, String message) {
        this.code = code;
        this.message = message;
    }
}
```
3. 定义一个错误响应类`ErrorResponse`：

```java
public class ErrorResponse extends BaseResponse {

    public ErrorResponse(IResponse response) {
        super(response.getCode(), response.getMessage());
    }
}
```

4. 定义一个异常基类`BaseException`，将 `IResponse` 作为其成员变量，这样，我们就将错误码信息包含到了异常类中：

```java
@Getter
public class BaseException extends RuntimeException {
    private final IResponse response;

    public BaseException(IResponse response) {
        this(response, null);
    }

    public BaseException(IResponse response, Throwable cause) {
        super(response.getMessage(), cause);
        this.response = response;
    }
}
```

5. 修改之前的`BusinessException`的实现方式，改成继承自：`BaseException`：

```java
@Getter
public class BusinessException extends BaseException {

    public BusinessException(IResponse response) {
        super(response);
    }

    public BusinessException(IResponse response, Throwable cause) {
        super(response, cause);
    }
}
```

6. 定义一个 Assert 接口，作为断言的基础接口，在接口实现了许多默认的方法：

```java
public interface Assert {
    /**
     * 创建异常
     *
     * @return BaseException 基础异常
     */
    BaseException newException();

    /**
     * 抛出异常
     */
    default void throwNewException() throws BaseException {
        throw newException();
    }

    /**
     * 创建异常
     *
     * @param cause 原因
     * @return BaseException 基础异常
     */
    BaseException newException(Throwable cause);

    /**
     * 抛出异常
     *
     * @param cause 原因
     */
    default void throwNewException(Throwable cause) throws BaseException {
        throw newException(cause);
    }

    /**
     * 断言条件为真，否则抛出异常
     *
     * @param condition 检查条件
     */
    default void assertTrue(boolean condition) {
        if (!condition) {
            throwNewException();
        }
    }

    /**
     * 断言条件为假，否则抛出异常
     *
     * @param condition 检查条件
     */
    default void assertFalse(boolean condition) {
        if (condition) {
            throwNewException();
        }
    }

    /**
     * 断言对象为空，否则抛出异常
     *
     * @param obj 检查的对象
     */
    default void assertNull(Object obj) {
        assertTrue(obj == null);
    }

    /**

CHANGES.md
Changelog
=========

v0.0.1 (20220303)
---------------------

- 初始版本

src/main/java/com/soulcraft/network/resp/HttpStatusEnum.java
package com.soulcraft.network.resp;

import com.soulcraft.network.exception.BusinessExceptionAssert;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.springframework.http.HttpStatus;

/**
 * <p>
 * HTTP返回信息枚举
 * </p>
 *
 * @author Scott
 * @see HttpStatus
 * @since 2022-03-10
 */
@Getter
@AllArgsConstructor
public enum HttpStatusEnum implements BusinessExceptionAssert {
    // 错误码搬运自：org.springframework.http.HttpStatus
    // 1xx Informational

    /**
     * {@code 100 Continue}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.2.1">HTTP/1.1: Semantics and Content, section 6.2.1</a>
     */
    CONTINUE(100, "Continue"),
    /**
     * {@code 101 Switching Protocols}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.2.2">HTTP/1.1: Semantics and Content, section 6.2.2</a>
     */
    SWITCHING_PROTOCOLS(101, "Switching Protocols"),
    /**
     * {@code 102 Processing}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc2518#section-10.1">WebDAV</a>
     */
    PROCESSING(102, "Processing"),
    /**
     * {@code 103 Checkpoint}.
     *
     * @see <a href="https://code.google.com/p/gears/wiki/ResumableHttpRequestsProposal">A proposal for supporting
     * resumable POST/PUT HTTP requests in HTTP/1.0</a>
     */
    CHECKPOINT(103, "Checkpoint"),

    // 2xx Success

    /**
     * {@code 200 OK}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.1">HTTP/1.1: Semantics and Content, section 6.3.1</a>
     */
    OK(200, "OK"),
    /**
     * {@code 201 Created}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.2">HTTP/1.1: Semantics and Content, section 6.3.2</a>
     */
    CREATED(201, "Created"),
    /**
     * {@code 202 Accepted}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.3">HTTP/1.1: Semantics and Content, section 6.3.3</a>
     */
    ACCEPTED(202, "Accepted"),
    /**
     * {@code 203 Non-Authoritative Information}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.4">HTTP/1.1: Semantics and Content, section 6.3.4</a>
     */
    NON_AUTHORITATIVE_INFORMATION(203, "Non-Authoritative Information"),
    /**
     * {@code 204 No Content}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.5">HTTP/1.1: Semantics and Content, section 6.3.5</a>
     */
    NO_CONTENT(204, "No Content"),
    /**
     * {@code 205 Reset Content}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.3.6">HTTP/1.1: Semantics and Content, section 6.3.6</a>
     */
    RESET_CONTENT(205, "Reset Content"),
    /**
     * {@code 206 Partial Content}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7233#section-4.1">HTTP/1.1: Range Requests, section 4.1</a>
     */
    PARTIAL_CONTENT(206, "Partial Content"),
    /**
     * {@code 207 Multi-Status}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc4918#section-13">WebDAV</a>
     */
    MULTI_STATUS(207, "Multi-Status"),
    /**
     * {@code 208 Already Reported}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc5842#section-7.1">WebDAV Binding Extensions</a>
     */
    ALREADY_REPORTED(208, "Already Reported"),
    /**
     * {@code 226 IM Used}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc3229#section-10.4.1">Delta encoding in HTTP</a>
     */
    IM_USED(226, "IM Used"),

    // 3xx Redirection

    /**
     * {@code 300 Multiple Choices}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.1">HTTP/1.1: Semantics and Content, section 6.4.1</a>
     */
    MULTIPLE_CHOICES(300, "Multiple Choices"),
    /**
     * {@code 301 Moved Permanently}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.2">HTTP/1.1: Semantics and Content, section 6.4.2</a>
     */
    MOVED_PERMANENTLY(301, "Moved Permanently"),
    /**
     * {@code 302 Found}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.3">HTTP/1.1: Semantics and Content, section 6.4.3</a>
     */
    FOUND(302, "Found"),
    /**
     * {@code 302 Moved Temporarily}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc1945#section-9.3">HTTP/1.0, section 9.3</a>
     * @deprecated in favor of {@link #FOUND} which will be returned from {@code HttpStatus.valueOf(302)}
     */
    @Deprecated
    MOVED_TEMPORARILY(302, "Moved Temporarily"),
    /**
     * {@code 303 See Other}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.4">HTTP/1.1: Semantics and Content, section 6.4.4</a>
     */
    SEE_OTHER(303, "See Other"),
    /**
     * {@code 304 Not Modified}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7232#section-4.1">HTTP/1.1: Conditional Requests, section 4.1</a>
     */
    NOT_MODIFIED(304, "Not Modified"),
    /**
     * {@code 305 Use Proxy}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.5">HTTP/1.1: Semantics and Content, section 6.4.5</a>
     * @deprecated due to security concerns regarding in-band configuration of a proxy
     */
    @Deprecated
    USE_PROXY(305, "Use Proxy"),
    /**
     * {@code 307 Temporary Redirect}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.4.7">HTTP/1.1: Semantics and Content, section 6.4.7</a>
     */
    TEMPORARY_REDIRECT(307, "Temporary Redirect"),
    /**
     * {@code 308 Permanent Redirect}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7238">RFC 7238</a>
     */
    PERMANENT_REDIRECT(308, "Permanent Redirect"),

    // --- 4xx Client Error ---

    /**
     * {@code 400 Bad Request}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.1">HTTP/1.1: Semantics and Content, section 6.5.1</a>
     */
    BAD_REQUEST(400, "Bad Request"),
    /**
     * {@code 401 Unauthorized}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7235#section-3.1">HTTP/1.1: Authentication, section 3.1</a>
     */
    UNAUTHORIZED(401, "Unauthorized"),
    /**
     * {@code 402 Payment Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.2">HTTP/1.1: Semantics and Content, section 6.5.2</a>
     */
    PAYMENT_REQUIRED(402, "Payment Required"),
    /**
     * {@code 403 Forbidden}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.3">HTTP/1.1: Semantics and Content, section 6.5.3</a>
     */
    FORBIDDEN(403, "Forbidden"),
    /**
     * {@code 404 Not Found}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.4">HTTP/1.1: Semantics and Content, section 6.5.4</a>
     */
    NOT_FOUND(404, "Not Found"),
    /**
     * {@code 405 Method Not Allowed}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.5">HTTP/1.1: Semantics and Content, section 6.5.5</a>
     */
    METHOD_NOT_ALLOWED(405, "Method Not Allowed"),
    /**
     * {@code 406 Not Acceptable}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.6">HTTP/1.1: Semantics and Content, section 6.5.6</a>
     */
    NOT_ACCEPTABLE(406, "Not Acceptable"),
    /**
     * {@code 407 Proxy Authentication Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7235#section-3.2">HTTP/1.1: Authentication, section 3.2</a>
     */
    PROXY_AUTHENTICATION_REQUIRED(407, "Proxy Authentication Required"),
    /**
     * {@code 408 Request Timeout}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.7">HTTP/1.1: Semantics and Content, section 6.5.7</a>
     */
    REQUEST_TIMEOUT(408, "Request Timeout"),
    /**
     * {@code 409 Conflict}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.8">HTTP/1.1: Semantics and Content, section 6.5.8</a>
     */
    CONFLICT(409, "Conflict"),
    /**
     * {@code 410 Gone}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.9">
     * HTTP/1.1: Semantics and Content, section 6.5.9</a>
     */
    GONE(410, "Gone"),
    /**
     * {@code 411 Length Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.10">
     * HTTP/1.1: Semantics and Content, section 6.5.10</a>
     */
    LENGTH_REQUIRED(411, "Length Required"),
    /**
     * {@code 412 Precondition failed}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7232#section-4.2">
     * HTTP/1.1: Conditional Requests, section 4.2</a>
     */
    PRECONDITION_FAILED(412, "Precondition Failed"),
    /**
     * {@code 413 Payload Too Large}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.11">
     * HTTP/1.1: Semantics and Content, section 6.5.11</a>
     * @since 4.1
     */
    PAYLOAD_TOO_LARGE(413, "Payload Too Large"),
    /**
     * {@code 413 Request Entity Too Large}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc2616#section-10.4.14">HTTP/1.1, section 10.4.14</a>
     * @deprecated in favor of {@link #PAYLOAD_TOO_LARGE} which will be
     * returned from {@code HttpStatus.valueOf(413)}
     */
    @Deprecated
    REQUEST_ENTITY_TOO_LARGE(413, "Request Entity Too Large"),
    /**
     * {@code 414 URI Too Long}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.12">
     * HTTP/1.1: Semantics and Content, section 6.5.12</a>
     * @since 4.1
     */
    URI_TOO_LONG(414, "URI Too Long"),
    /**
     * {@code 414 Request-URI Too Long}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc2616#section-10.4.15">HTTP/1.1, section 10.4.15</a>
     * @deprecated in favor of {@link #URI_TOO_LONG} which will be returned from {@code HttpStatus.valueOf(414)}
     */
    @Deprecated
    REQUEST_URI_TOO_LONG(414, "Request-URI Too Long"),
    /**
     * {@code 415 Unsupported Media Type}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.13">
     * HTTP/1.1: Semantics and Content, section 6.5.13</a>
     */
    UNSUPPORTED_MEDIA_TYPE(415, "Unsupported Media Type"),
    /**
     * {@code 416 Requested Range Not Satisfiable}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7233#section-4.4">HTTP/1.1: Range Requests, section 4.4</a>
     */
    REQUESTED_RANGE_NOT_SATISFIABLE(416, "Requested range not satisfiable"),
    /**
     * {@code 417 Expectation Failed}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.5.14">
     * HTTP/1.1: Semantics and Content, section 6.5.14</a>
     */
    EXPECTATION_FAILED(417, "Expectation Failed"),
    /**
     * {@code 418 I'm a teapot}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc2324#section-2.3.2">HTCPCP/1.0</a>
     */
    I_AM_A_TEAPOT(418, "I'm a teapot"),
    /**
     * @deprecated See
     * <a href="https://tools.ietf.org/rfcdiff?difftype=--hwdiff&url2=draft-ietf-webdav-protocol-06.txt">
     * WebDAV Draft Changes</a>
     */
    @Deprecated
    INSUFFICIENT_SPACE_ON_RESOURCE(419, "Insufficient Space On Resource"),
    /**
     * @deprecated See
     * <a href="https://tools.ietf.org/rfcdiff?difftype=--hwdiff&url2=draft-ietf-webdav-protocol-06.txt">
     * WebDAV Draft Changes</a>
     */
    @Deprecated
    METHOD_FAILURE(420, "Method Failure"),
    /**
     * @deprecated See <a href="https://tools.ietf.org/rfcdiff?difftype=--hwdiff&url2=draft-ietf-webdav-protocol-06.txt">
     * WebDAV Draft Changes</a>
     */
    @Deprecated
    DESTINATION_LOCKED(421, "Destination Locked"),
    /**
     * {@code 422 Unprocessable Entity}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc4918#section-11.2">WebDAV</a>
     */
    UNPROCESSABLE_ENTITY(422, "Unprocessable Entity"),
    /**
     * {@code 423 Locked}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc4918#section-11.3">WebDAV</a>
     */
    LOCKED(423, "Locked"),
    /**
     * {@code 424 Failed Dependency}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc4918#section-11.4">WebDAV</a>
     */
    FAILED_DEPENDENCY(424, "Failed Dependency"),
    /**
     * {@code 425 Too Early}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc8470">RFC 8470</a>
     * @since 5.2
     */
    TOO_EARLY(425, "Too Early"),
    /**
     * {@code 426 Upgrade Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc2817#section-6">Upgrading to TLS Within HTTP/1.1</a>
     */
    UPGRADE_REQUIRED(426, "Upgrade Required"),
    /**
     * {@code 428 Precondition Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc6585#section-3">Additional HTTP Status Codes</a>
     */
    PRECONDITION_REQUIRED(428, "Precondition Required"),
    /**
     * {@code 429 Too Many Requests}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc6585#section-4">Additional HTTP Status Codes</a>
     */
    TOO_MANY_REQUESTS(429, "Too Many Requests"),
    /**
     * {@code 431 Request Header Fields Too Large}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc6585#section-5">Additional HTTP Status Codes</a>
     */
    REQUEST_HEADER_FIELDS_TOO_LARGE(431, "Request Header Fields Too Large"),
    /**
     * {@code 451 Unavailable For Legal Reasons}.
     *
     * @see <a href="https://tools.ietf.org/html/draft-ietf-httpbis-legally-restricted-status-04">
     * An HTTP Status Code to Report Legal Obstacles</a>
     * @since 4.3
     */
    UNAVAILABLE_FOR_LEGAL_REASONS(451, "Unavailable For Legal Reasons"),

    // --- 5xx Server Error ---

    /**
     * {@code 500 Internal Server Error}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.1">HTTP/1.1: Semantics and Content, section 6.6.1</a>
     */
    INTERNAL_SERVER_ERROR(500, "Internal Server Error"),
    /**
     * {@code 501 Not Implemented}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.2">HTTP/1.1: Semantics and Content, section 6.6.2</a>
     */
    NOT_IMPLEMENTED(501, "Not Implemented"),
    /**
     * {@code 502 Bad Gateway}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.3">HTTP/1.1: Semantics and Content, section 6.6.3</a>
     */
    BAD_GATEWAY(502, "Bad Gateway"),
    /**
     * {@code 503 Service Unavailable}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.4">HTTP/1.1: Semantics and Content, section 6.6.4</a>
     */
    SERVICE_UNAVAILABLE(503, "Service Unavailable"),
    /**
     * {@code 504 Gateway Timeout}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.5">HTTP/1.1: Semantics and Content, section 6.6.5</a>
     */
    GATEWAY_TIMEOUT(504, "Gateway Timeout"),
    /**
     * {@code 505 HTTP Version Not Supported}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc7231#section-6.6.6">HTTP/1.1: Semantics and Content, section 6.6.6</a>
     */
    HTTP_VERSION_NOT_SUPPORTED(505, "HTTP Version not supported"),
    /**
     * {@code 506 Variant Also Negotiates}
     *
     * @see <a href="https://tools.ietf.org/html/rfc2295#section-8.1">Transparent Content Negotiation</a>
     */
    VARIANT_ALSO_NEGOTIATES(506, "Variant Also Negotiates"),
    /**
     * {@code 507 Insufficient Storage}
     *
     * @see <a href="https://tools.ietf.org/html/rfc4918#section-11.5">WebDAV</a>
     */
    INSUFFICIENT_STORAGE(507, "Insufficient Storage"),
    /**
     * {@code 508 Loop Detected}
     *
     * @see <a href="https://tools.ietf.org/html/rfc5842#section-7.2">WebDAV Binding Extensions</a>
     */
    LOOP_DETECTED(508, "Loop Detected"),
    /**
     * {@code 509 Bandwidth Limit Exceeded}
     */
    BANDWIDTH_LIMIT_EXCEEDED(509, "Bandwidth Limit Exceeded"),
    /**
     * {@code 510 Not Extended}
     *
     * @see <a href="https://tools.ietf.org/html/rfc2774#section-7">HTTP Extension Framework</a>
     */
    NOT_EXTENDED(510, "Not Extended"),
    /**
     * {@code 511 Network Authentication Required}.
     *
     * @see <a href="https://tools.ietf.org/html/rfc6585#section-6">Additional HTTP Status Codes</a>
     */
    NETWORK_AUTHENTICATION_REQUIRED(511, "Network Authentication Required"),
    ;

    /**
     * 返回码
     */
    private int code;
    /**
     * 返回消息
     */
    private String message;

    public String getAppName() {
        return "COM";
    }

    public String getModuleName() {
        return "HTTP";
    }

}
